core/modules/file/file.field.inc

```diff
@@ -789,14 +798,55 @@ function file_field_widget_value($element, $input = FALSE, $form_state) {
 
   // Ensure that all the required properties are returned even if empty.
   $return += array(
-    'fid' => 0,
+    'fid' => array(),
     'display' => 1,
     'description' => '',
   );
 
   return $return;
 }
 
+ /**
+ * Validation callback for upload element on file widget. Checks if user has
+ * uploaded more files than allowed.
+ *
+ * This validator is used only when cardinality not set to 1 or unlimited.
+ */
+function file_field_widget_multiple_count_validate($element, &$form_state, $form) {
+  $parents = $element['#parents'];
+  $values = backdrop_array_get_nested_value($form_state['values'], $parents);
+
+  array_pop($parents);
+  $current = count(element_children(backdrop_array_get_nested_value($form, $parents))) - 1;
+
+  $field = field_info_field($element['#field_name']);
+  $uploaded = count($values['fid']);
+  $count = $uploaded + $current;
+  if ($count > $field['cardinality']) {
+    $keep = $uploaded - $count + $field['cardinality'];
+    $removed_files = array_slice($values['fid'], $keep);
+    $removed_names = array();
+    foreach ($removed_files as $fid) {
+      $file = file_load($fid);
+      $removed_names[] = $file->filename;
+    }
+    backdrop_set_message(
+      t(
+        'Field %field can only hold @max values but there were @count uploaded. The following files have been omitted as a result: %list.',
+        array(
+          '%field' => $field['field_name'],
+          '@max' => $field['cardinality'],
+          '@count' => $keep,
+          '%list' => implode(', ', $removed_names),
```

Comment:
Perhaps use `theme_item_list` here for better readability in case of many removed files.